client/apiversion.go

```diff
@@ -0,0 +1,183 @@
+	Links   []apiVersionLink `json:"links"`
+	Status  string           `json:"status"`
+}
+
+type apiVersionLink struct {
+	Href string `json:"href"`
+	Rel  string `json:"rel"`
+}
+
+type apiURLVersion struct {
+	rootURL          url.URL
+	serviceURLSuffix string
+	versions         []apiVersionInfo
+}
+
+// Return a full formed serviceURL based on the API version requested, the rootURL and
+// the serviceURLSuffix.  If not match to the requested API version an error is returned.
+// If only the major version is defined for the requested version, the highest minor
+// version available is assumed.
+func getAPIVersionURL(apiURLVersionInfo *apiURLVersion, requested apiVersion) (string, error) {
+	var match string
+	for _, v := range apiURLVersionInfo.versions {
+		if v.Version.major != requested.major {
+			continue
+		}
+		if requested.minor != -1 && v.Version.minor != requested.minor {
+			continue
+		}
+		for _, link := range v.Links {
+			if link.Rel != "self" {
+				continue
+			}
+			hrefURL, err := url.Parse(link.Href)
+			if err != nil {
+				return "", err
+			}
+			match = hrefURL.Path
+		}
+		if requested.minor != -1 {
+			break
+		}
+	}
+	if match == "" {
+		return "", fmt.Errorf("could not find matching URL")
+	}
+	versionURL := apiURLVersionInfo.rootURL
+	versionURL.Path = path.Join(versionURL.Path, match, apiURLVersionInfo.serviceURLSuffix)
+	return versionURL.String(), nil
+}
+
+func (v *apiVersion) UnmarshalJSON(b []byte) error {
+	var s string
+	if err := json.Unmarshal(b, &s); err != nil {
+		return err
+	}
+	parsed, err := parseVersion(s)
+	if err != nil {
+		return err
+	}
+	*v = parsed
+	return nil
+}
+
+func parseVersion(s string) (apiVersion, error) {
+	s = strings.TrimPrefix(s, "v")
+	parts := strings.SplitN(s, ".", 2)
+	if len(parts) == 0 || len(parts) > 2 {
+		return apiVersion{}, fmt.Errorf("invalid API version %q", s)
+	}
+	var minor int = -1
+	major, err := strconv.Atoi(parts[0])
+	if err != nil {
+		return apiVersion{}, err
+	}
+	if len(parts) == 2 {
+		var err error
+		minor, err = strconv.Atoi(parts[1])
+		if err != nil {
+			return apiVersion{}, err
+		}
+	}
+	return apiVersion{major, minor}, nil
+}
+
+func unmarshallVersion(Versions json.RawMessage) ([]apiVersionInfo, error) {
+	// Some services respond with {"versions":[...]}, and
+	// some respond with {"versions":{"values":[...]}}.
+	var object interface{}
+	var versions []apiVersionInfo
+	if err := json.Unmarshal(Versions, &object); err != nil {
+		return versions, err
+	}
+	if _, ok := object.(map[string]interface{}); ok {
+		var valuesObject struct {
+			Values []apiVersionInfo `json:"values"`
+		}
+		if err := json.Unmarshal(Versions, &valuesObject); err != nil {
+			return versions, err
+		}
+		versions = valuesObject.Values
+	} else {
+		if err := json.Unmarshal(Versions, &versions); err != nil {
+			return versions, err
+		}
+	}
+	return versions, nil
+}
+
+// Given a service endpoint, return data on the versions suported by the service
```

Comment:
sorry to be picky - the comment should start with the method name and end with a period. eg
// getAPIVersions returns data on the versions supported by the service at the specified endpoint.

Same comment applies to the getAPIVersionURL method above